git_machete/git_operations.py

```diff
@@ -0,0 +1,378 @@
+from typing import List, Optional, Callable, Tuple, Dict, Match, Set
+
+import os
+import re
+import sys
+
+from git_machete.contexts import CommandLineContext
+from git_machete.exceptions import MacheteException
+from git_machete import utils
+
+REFLOG_ENTRY = Tuple[str, str]
+
+
+class GitContext:
+
+    counterparts_for_fetching_cached: Optional[Dict[str, Optional[str]]] = None  # TODO (#110): default dict with None
+    git_version: Tuple[int, ...] = None
+    root_dir: str = None
+    git_dir: str = None
+    config_cached: Optional[Dict[str, str]] = None
+    remotes_cached: Optional[List[str]] = None
+    fetch_done_for: Set[str] = set()
+    short_commit_sha_by_revision_cached: Dict[str, str] = {}
+    tree_sha_by_commit_sha_cached: Optional[Dict[str, Optional[str]]] = None  # TODO (#110): default dict with None
+    commit_sha_by_revision_cached: Optional[Dict[str, Optional[str]]] = None  # TODO (#110): default dict with None
+    committer_unix_timestamp_by_revision_cached: Optional[Dict[str, int]] = None  # TODO (#110): default dict with 0
+    local_branches_cached: Optional[List[str]] = None
+    remote_branches_cached: Optional[List[str]] = None
+    initial_log_shas_cached: Dict[str, List[str]] = {}
+    remaining_log_shas_cached: Dict[str, List[str]] = {}
+    reflogs_cached: Optional[Dict[str, Optional[List[REFLOG_ENTRY]]]] = None
+
+    def __init__(self) -> None:
+        pass
+
+
+def run_git(cli_ctxt: CommandLineContext, git_cmd: str, *args: str, **kwargs: Dict[str, str]) -> int:
```

Comment:
Hmmm would it make sense to, instead of keeping all these function global, convert them into instance methods of `GitContext`?

Reply:
This way, instead of keeping things like `config_cached` as class attributes (i.e. de facto still **global** variables), you could move them into instance attributes of `GitContext`